src/array/utf8_array.rs

```diff
@@ -1,28 +1,64 @@
 // Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
 
 use std::iter::FromIterator;
+use std::marker::PhantomData;
 
 use bitvec::vec::BitVec;
 use serde::{Deserialize, Serialize};
 
 use super::{Array, ArrayBuilder, ArrayEstimateExt, ArrayValidExt};
+use crate::types::BlobRef;
 
-/// A collection of Rust UTF8 `String`s.
-#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct Utf8Array {
+/// A collection of variable-length values.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct BytesArray<T: ValueRef + ?Sized> {
     offset: Vec<usize>,
     valid: BitVec,
     data: Vec<u8>,
+    _type: PhantomData<T>,
 }
 
-impl Array for Utf8Array {
-    type Item = str;
-    type Builder = Utf8ArrayBuilder;
+/// The borrowed type of a variable-length value.
+pub trait ValueRef: ToOwned + AsRef<[u8]> + Send + Sync + 'static {
+    fn from_bytes(s: &[u8]) -> &Self;
+}
+
+impl ValueRef for str {
+    fn from_bytes(s: &[u8]) -> &Self {
+        unsafe { std::str::from_utf8_unchecked(s) }
+    }
+}
+
+impl ValueRef for BlobRef {
```

Comment:
Can we implement `ValueRef` directly on `[u8]`, so that we can get rid of the unsafe transmute between `BlobRef` and `&[u8]`?

Reply:
Per discussion at #381, what if in the future we want to store something like `blob_type` in `Blob`? If we define `BlobRef` as:

```
pub struct BlobRef(BlobType, [u8]);
```

Can we still do the transmute like this?

Reply:
> Can we implement `ValueRef` directly on `[u8]`, so that we can get rid of the unsafe transmute between `BlobRef` and `&[u8]`?

Yes, but this way has some limitations, such as `&[u8]` doesn't implement `Display` which will bring trouble to compilation. We need to customize the output format on the new type. And the unsafe transmute from `&[u8]` to `&T: ValueRef` is unavoidable, because not all bytes are valid `T`. Only the bytes in array can be safely converted.

> Per discussion at #381, what if in the future we want to store something like blob_type in Blob? If we define BlobRef as:
> ```
> pub struct BlobRef(BlobType, [u8]);
> ```
> Can we still do the transmute like this?

Nope, the new type must be defined exactly as `[u8]`, in order to have the same memory layout. This is the limitation of this approach. 
If we need something like `blob_type`, we should encode it into the `[u8]`. For example we can encode the first byte as type and the rest as content, then the code will be:
```rust
pub struct BlobRef([u8]);

impl BlobRef {
    pub fn type_(&self) -> u8 {
        self.0[0]
    }
    pub fn content(&self) -> &[u8] {
        &self.0[1..]
    }
}
```